ttnn/cpp/ttnn/operations/data_movement/repeat/repeat.cpp

```diff
@@ -33,14 +49,59 @@ ttnn::Tensor RepeatOperation::invoke(
+                auto outputs = operation::run_without_autoformat(RepeatDeviceOperation{dim, repeat_dims[dim], memory_config}, {output});
+                TT_FATAL(outputs.size() == 1, "ttnn.repeat: expected 1 output tensor from run_without_autoformat, but got {}", outputs.size());
+                output = outputs[0];
             }
             return {output};
-        }, {input_tensor}, output_tensors);
-    return output_tensors.at(0);
+        }, {}, output_tensors);
+    TT_FATAL(output_tensors.size() == 1, "ttnn.repeat: expected 1 output tensor, but got {}", output_tensors.size());
+    if (input_tensor.get_layout() != Layout::ROW_MAJOR
+        && logical_input_shape != padded_input_shape) {
+        auto zero_indices = std::vector<uint32_t>(input_rank, 0);
+        std::vector<uint32_t> end_indices(input_rank);
+        for (uint32_t i = 0; i < input_rank; ++i) {
+            end_indices[i] = repeated_logical_shape[i];
+        }
+        auto step = std::vector<uint32_t>(input_rank, 1);
+
+        if (repeated_logical_shape.volume() % tt::constants::TILE_HW != 0) {
+            // volume of the repeated tensor doesn't fit neatly into tiles.
+            // slice doesn't support padding on the output for now, so we need
+            // to perform the slice in row-major then re-tilize ourselves.
+            auto rm_output = ttnn::untilize(output_tensors[0]);
+            auto sliced_output = ttnn::slice(rm_output, zero_indices, end_indices, step, input_tensor.memory_config(), std::nullopt);
+            auto sliced_shape = sliced_output.get_shape();
+
+            auto padded_height = tt::round_up(sliced_shape[-2], tt::constants::TILE_HEIGHT);
+            auto padded_width = tt::round_up(sliced_shape[-1], tt::constants::TILE_WIDTH);
+            auto padding_vec = std::vector<std::pair<uint32_t, uint32_t>>(input_rank-2, std::pair<uint32_t, uint32_t>(0, 0));
+            padding_vec.reserve(input_rank);
+            padding_vec.push_back(std::pair<uint32_t, uint32_t>(0, padded_height - sliced_output.get_padded_shape()[-2]));
+            padding_vec.push_back(std::pair<uint32_t, uint32_t>(0, padded_width - sliced_output.get_padded_shape()[-1]));
+
+            constexpr bool pad_use_multicore = true;
+            sliced_output = ttnn::pad(queue_id, sliced_output, padding_vec, 0.0f, pad_use_multicore, std::nullopt);
+            sliced_output = ttnn::tilize(sliced_output, input_tensor.memory_config());
+
+            auto sliced_shape_vec = std::vector<uint32_t>(input_rank,0);
+            auto padded_shape_vec = std::vector<uint32_t>(input_rank,0);
+            for (uint32_t i = 0; i < input_rank; ++i) {
+                padded_shape_vec[i] = sliced_output.get_padded_shape()[i];
+                sliced_shape_vec[i] = sliced_shape[i];
+            }
+
+            auto padded_shape = ttnn::types::Shape(sliced_shape_vec, padded_shape_vec);
```

Comment:
is this the same shape as `ttnn::Shape`? If yes, could you please make it ttnn::Shape (this way it will be easier for people who plan to refactor ttnn::Shape to find it)